src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -105,32 +111,44 @@ internal static unsafe bool IsValid(char* name, int start, ref int end, bool all
         //                 / "2" %x30-34 DIGIT     ; 200-249
         //                 / "25" %x30-35          ; 250-255
         //
-        internal static unsafe bool IsValidCanonical(char* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+        internal static unsafe bool IsValidCanonical<TChar>(TChar* name, int start, ref int end, bool allowIPv6, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
             int dots = 0;
             int number = 0;
             bool haveNumber = false;
             bool firstCharIsZero = false;
 
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
             while (start < end)
             {
-                char ch = name[start];
+                TChar ch = name[start];
+
                 if (allowIPv6)
                 {
-                    // for ipv4 inside ipv6 the terminator is either ScopeId, prefix or ipv6 terminator
-                    if (ch == ']' || ch == '/' || ch == '%')
+                    // For an IPv4 address nested inside an IPv6, the terminator is either ScopeId ('%'), prefix ('/') or ipv6 address terminator (']')
+                    if (ch == TChar.CreateTruncating('%') || ch == TChar.CreateTruncating('/') || ch == TChar.CreateTruncating(']'))
+                    {
                         break;
+                    }
                 }
-                else if (ch == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
+                // For a normal IPv4 address, the terminator is the prefix ('/' or its counterpart, '\'). If notImplicitFile is set, the terminator
+                // is one of the characters which signify the start of the rest of the URI - the port number (':'), query string ('?') or fragment ('#')
```

Comment:
We try to avoid having comments like these between if/else blocks. You can move it to be inside the else if, effectively commenting on the break.